smartlearner/interfaces/loss.py

```diff
@@ -5,20 +6,29 @@ class Loss(object):
     def __init__(self, model, dataset):
         self.model = model
         self.dataset = dataset
-        self.target = dataset.symb_targets
         self.consider_constant = []  # Part of the computational graph to be considered as a constant.
+        self.graph_updates = OrderedDict()
+        self._gradients = None
+
+    @property
+    def target(self):
+        return self.dataset.symb_targets
 
     def get_graph_output(self):
         output, updates = self.model.get_model_output(self.dataset.symb_inputs)
         return self._loss_function(output), updates
 
-    def get_gradients(self):
-        cost, updates = self.get_graph_output()
-        gparams = T.grad(cost=cost,
-                         wrt=self.model.parameters,
-                         consider_constant=self.consider_constant)
-        gradients = dict(zip(self.model.parameters, gparams))
-        return gradients, updates
+    @property
+    def gradients(self):
+        if self._gradients is None:
+            cost, updates_from_model_output = self.get_graph_output()
+            gparams = T.grad(cost=cost,
+                             wrt=self.model.parameters,
+                             consider_constant=self.consider_constant)
+            self._gradients = dict(zip(self.model.parameters, gparams))
+            self.graph_updates.update(updates_from_model_output)
+
+        return self._gradients
```

Comment:
Hmm, I'm thinking their might be a way to keep how it was (i.e. returning the gradients and the updates). We only need to move this line `updates = self._optimizer.gather_updates()`, which is currently in the method `trainer._build_theano_graph` into the constructor of `Trainer`! That way we can guarantee that the entire graph is  built once the trainer object is created. After that, the only things an user should do is to add tasks. @ASalvail what do you think of this approach?